flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannelTest.java

```diff
@@ -301,81 +306,388 @@ public void testProducerFailedException() throws Exception {
+			inputChannel.releaseAllResources();
+
+			networkBufferPool.destroyAllBufferPools();
+			networkBufferPool.destroy();
+
+			executor.shutdown();
 		}
-		inputChannel.assignExclusiveSegments(exclusiveSegments);
+	}
 
-		assertEquals("The number of available buffers is not equal to the assigned amount.",
-			numExclusiveBuffers, inputChannel.getNumberOfAvailableBuffers());
+	/**
+	 * Tests to verify that there is no race condition with two things running in parallel:
+	 * recycling the exclusive or floating buffers and some other thread releasing the
+	 * input channel.
+	 */
+	@Test
+	public void testConcurrentRecycleAndRelease() throws Exception {
+		// Setup
+		final NetworkBufferPool networkBufferPool = new NetworkBufferPool(256, 32, MemoryType.HEAP);
+		final ExecutorService executor = Executors.newFixedThreadPool(2);
+		final SingleInputGate inputGate = createSingleInputGate();
+		final RemoteInputChannel inputChannel  = createRemoteInputChannel(inputGate);
+		inputGate.setInputChannel(inputChannel.partitionId.getPartitionId(), inputChannel);
+		try {
+			final int numFloatingBuffers = 128;
+			final int numExclusiveSegments = 2;
+			final BufferPool bufferPool = networkBufferPool.createBufferPool(numFloatingBuffers, numFloatingBuffers);
+			inputGate.setBufferPool(bufferPool);
+			inputGate.assignExclusiveSegments(networkBufferPool, numExclusiveSegments);
+
+			// Exhaust all the floating buffers
+			final List<Buffer> floatingBuffers = new ArrayList<>(numFloatingBuffers);
+			for (int i = 0; i < numFloatingBuffers; i++) {
+				Buffer buffer = bufferPool.requestBuffer();
+				assertNotNull(buffer);
+				floatingBuffers.add(buffer);
+			}
+
+			// Exhaust all the exclusive buffers
+			final List<Buffer> exclusiveBuffers = new ArrayList<>(numExclusiveSegments);
+			for (int i = 0; i < numExclusiveSegments; i++) {
+				Buffer buffer = inputChannel.requestBuffer();
+				assertNotNull(buffer);
+				exclusiveBuffers.add(buffer);
+			}
+
+			final Callable<Void> recycleBufferTask = new Callable<Void>() {
+				@Override
+				public Void call() throws Exception {
+					// Recycle all the exclusive buffers
+					for (Buffer buffer : exclusiveBuffers) {
+						buffer.recycle();
+					}
+
+					// Recycle all the floating buffers
+					for (Buffer buffer : floatingBuffers) {
+						buffer.recycle();
+					}
+
+					return null;
+				}
+			};
+
+			final Callable<Void> releaseTask = new Callable<Void>() {
+				@Override
+				public Void call() throws Exception {
+					inputChannel.releaseAllResources();
+
+					return null;
+				}
+			};
+
+			// Submit tasks and wait to finish
+			final List<Future<Void>> results = Lists.newArrayListWithCapacity(2);
+			results.add(executor.submit(recycleBufferTask));
+			results.add(executor.submit(releaseTask));
+			for (Future<Void> result : results) {
+				result.get();
+			}
+
+			assertEquals("There should be no buffers available in the channel.",
+				0, inputChannel.getNumberOfAvailableBuffers());
```

Comment:
also verify the `bufferPool.getNumberOfAvailableMemorySegments()` and `networkBufferPool.getNumberOfAvailableMemorySegments`?

Reply:
When the channel is released by one thread, we are not sure whether the floating buffer is requested and how many are requested by another thread.

If the number of requested floating buffers before is less than the capacity of pool, it can not verify how many available floating buffers currently in the pool after channel released, and the same as `NetworkBufferPool`.